src/BenchmarkDotNet/Validators/ExecutionValidatorBase.cs

```diff
@@ -121,135 +146,115 @@ private bool TryToCallGlobalMethod<T>(object benchmarkTypeInstance, List<Validat
             }
 
-            foreach (var paramField in paramFields)
+            var property = task.GetType().GetProperty("Result", BindingFlags.Public | BindingFlags.Instance);
+            if (property is null)
             {
-                if (!paramField.IsPublic)
-                {
-                    errors.Add(new ValidationError(
-                        TreatsWarningsAsErrors,
-                        $"Fields marked with [Params] must be public, {paramField.Name} of {benchmarkTypeInstance.GetType().Name} is not"));
-
-                    return false;
-                }
-
-                var values = paramField.GetCustomAttributes(false).OfType<ParamsAttribute>().Single().Values;
-                if (!values.Any())
-                {
-                    errors.Add(new ValidationError(
-                        TreatsWarningsAsErrors,
-                        $"Fields marked with [Params] must have some values defined, {paramField.Name} of {benchmarkTypeInstance.GetType().Name} has none"));
-
-                    return false;
-                }
-
-                try
-                {
-                    paramField.SetValue(benchmarkTypeInstance, values.First());
-                }
-                catch (Exception ex)
-                {
-                    errors.Add(new ValidationError(
-                        TreatsWarningsAsErrors,
-                        $"Failed to set {paramField.Name} of {benchmarkTypeInstance.GetType().Name} to {values.First()}, exception was: {GetDisplayExceptionMessage(ex)}"));
-
-                    return false;
-                }
+                return false;
             }
 
+            result = property.GetValue(task);
             return true;
         }
 
-        private bool TryToSetParamsProperties(object benchmarkTypeInstance, List<ValidationError> errors)
+        private bool TryToFillParameters(BenchmarkCase benchmark, object benchmarkTypeInstance, List<ValidationError> errors)
         {
-            var paramProperties = benchmarkTypeInstance
-                .GetType()
-                .GetAllProperties()
-                .Where(propertyInfo => propertyInfo.GetCustomAttributes(false).OfType<ParamsAttribute>().Any())
-                .ToArray();
+            if (ValidateMembers<ParamsAttribute>(benchmarkTypeInstance, errors))
+                return false;
 
-            if (!paramProperties.Any())
-            {
-                return true;
-            }
+            if (ValidateMembers<ParamsSourceAttribute>(benchmarkTypeInstance, errors))
+                return false;
+
+            bool hasError = false;
 
-            foreach (var paramProperty in paramProperties)
+            foreach (var parameter in benchmark.Parameters.Items)
             {
-                var setter = paramProperty.SetMethod;
-                if (setter == null || !setter.IsPublic)
+                // InProcessNoEmitToolchain does not support arguments
+                if (!parameter.IsArgument)
```

Comment:
Can that be added as a `ValidationError` if the toolchain is `InProcessNoEmitToolchain`?

Reply:
I did not understand your question.
`ReturnValueValidator`/`ExecutionValidator` works with `InProcessToolchain` only under the hood.

Reply:
I mean, you're checking for that condition already, so if the user is actually using `InProcessNoEmitToolchain`, that should add the `ValidationError` there instead of continuing. Otherwise I guess it will fail downstream when the `BenchmarkActionFactory` actually builds the actions.

Reply:
BDN treats parameters and arguments as one entity. So, we should split them:

I only pass parameters to `InProcessNoEmitRunner.FillMember` because arguments are not supported.
I'm passing arguments here (they are already filled with values):
https://github.com/dotnet/BenchmarkDotNet/blob/ab54ca823490b49d42999f4a701f4c630e68770e/src/BenchmarkDotNet/Validators/ExecutionValidatorBase.cs#L286

Now I understand you correctly? 😊

Reply:
Oh I guess this makes more sense to do in the `BenchmarkExecutor` than here. Something like this:

```cs
if (arguments.Length > 0)
{
    foreach (var job in benchmark.Config.GetJobs())
    {
#pragma warning disable CS0618 // Type or member is obsolete
        if (job.Infrastructure.Toolchain is InProcessNoEmitToolchain || job.Infrastructure.Toolchain is InProcessToolchain)
#pragma warning restore CS0618 // Type or member is obsolete
        {
            errors.Add(new ValidationError(
                true,
                $"Failed to execute benchmark '{benchmark.DisplayInfo}' with {job.Infrastructure.Toolchain.GetType().Name}, this toolchain does not support arguments.",
                benchmark));
        }
    }
}
```

Reply:
I'm not sure that those optional validators are supposed to validate toolchains.

I think this is a job for the new mandatory `ToolchainValidator`.